src/jit/codegenxarch.cpp

```diff
@@ -7377,14 +7377,48 @@ CodeGen::genCkfinite(GenTreePtr treeNode)
     assert(treeNode->gtRsvdRegs != RBM_NONE);
     assert(genCountBits(treeNode->gtRsvdRegs) == 1);
     regNumber tmpReg = genRegNumFromMask(treeNode->gtRsvdRegs);  
-    
+
+#ifdef _TARGET_AMD64_
+    // Copy the floating-point value to an integer register. If we copied a float to a long, then
+    // right-shift the value so the high 32 bits of the floating-point value sit in the low 32
+    // bits of the integer register.
     instruction ins = ins_CopyFloatToInt(targetType, (targetType == TYP_FLOAT) ? TYP_INT : TYP_LONG);
     inst_RV_RV(ins, genConsumeReg(op1), tmpReg, targetType);
     if (targetType == TYP_DOUBLE)
     {
         // right shift by 32 bits to get to exponent.
         inst_RV_SH(INS_shr, EA_8BYTE, tmpReg, 32);
     }
+#else // !_TARGET_AMD64_
+    // If the target type is TYP_DOUBLE, we want to extract the high 32 bits into the register.
+    // There is no easy way to do this. To not require an extra register, we'll use shuffles
+    // to move the high 32 bits into the low 32 bits, then then shuffle it back, since we
+    // need to produce the value into the target register.
+    //
+    // For TYP_DOUBLE, we'll generate:
+    //    if (op1reg != targetReg)
+    //    {
+    //        movaps targetReg, op1Reg
+    //    }
+    //    shufps targetReg, targetReg, 0xB1	// WZYX => ZWXY
+    //    mov_xmm2i tmpReg, targetReg		// tmpReg <= Y
+    //    and tmpReg, <mask>
+    //    cmp tmpReg, <mask>
+    //    je <throw block>
+    //    shufps targetReg, targetReg, 0xB1	// ZWXY => WZYX
+
+    genConsumeReg(op1);
+    regNumber targetReg = treeNode->gtRegNum;
+    if (targetType == TYP_DOUBLE)
+    {
+        if (treeNode->gtRegNum != op1->gtRegNum)
+        {
+            inst_RV_RV(ins_Copy(targetType), targetReg, op1->gtRegNum, targetType);
+        }
+        inst_RV_RV_IV(INS_shufps, EA_16BYTE, targetReg, targetReg, 0xb1);
+    }
+    inst_RV_RV(ins_CopyFloatToInt(TYP_FLOAT, TYP_INT), op1->gtRegNum, tmpReg, TYP_FLOAT);
+#endif // !_TARGET_AMD64_
 
```

Comment:
A couple of minor suggestions:
1. It would be nice to have a comment outside the #ifdefs that talks about what needs to be done (if the exponent is all 1's, throw ArithmeticException, otherwise copy the value to targetReg. Then, in the #ifdefs you can talk about how that's done in each case.
2. It would also be nice to extract the common parts (getting the targetReg, calling genConsumeReg and ins_CopyFloatToInt).